### PR TITLE
Add support for multiple schemas to dbDoc command

### PR DESCRIPTION
**Pull request by zpanderson**
## Environment
**Liquibase Version**: all

**Liquibase Integration & Version**: all

**Liquibase Extension(s) & Version**: all

**Database Vendor & Version**: all

**Operating System Type & Version**: all

## Pull Request Type

* [ ] Bug fix (non-breaking change which fixes an issue.)
* [x] Enhancement/New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
The current dbDoc implementation only allows for a single schema to be documented at a time which makes it hard to use the built in tools for implementations with multiple schemas managed by Liquibase. This enhancement allows the user to optionally specify a list of schemas to be included which mainly affects the Current Tables frame by adding collapsible (no javascript) sections for each schema. In the case that no schemas are provided the existing default behavior of using the default schema remains. 

## Steps To Reproduce
* Run the dbDoc command without the optional schemas parameter and the default schema should be used for the current tables frame.
* Run the dbDoc command with multiple schemas in the optional schemas parameter and there should now be multiple collapsible sections under the current tables frame. 

## Actual Behavior
* Only the default schema is ever considered for the dbDoc command

## Expected/Desired Behavior
* A list of schemas is able to be provided to the dbDoc command to allow support for multiple schemas.

## Additional Context
Was not sure if it's ok to re-use the optional schemas parameter which is already used by the 'Diff' command - left a placeholder question in the code that can be removed if this is OK to do.

## Fast Track PR Acceptance Checklist:

* [x] Build is successful and all new and existing tests pass
* [ ] Added [Unit Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)
* [ ] Added [Integration Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)
* [ ] Added [Test Harness Test(s)](https://github.com/liquibase/liquibase-test-harness/pulls)
* [ ] Documentation Updated

## Need Help?
Come chat with us on our [discord channel](https://discord.com/channels/700506481111597066/700506481572839505)

### CODE REVIEW

File: liquibase-core/src/main/java/liquibase/integration/commandline/Main.java [hunk 1]
```diff
@@ -1767,13 +1768,27 @@ protected void doMigration() throws Exception {
                 Scope.getCurrentScope().getUI().sendMessage(checkSum.toString());
                 return;
             } else if (COMMANDS.DB_DOC.equalsIgnoreCase(command)) {
+                // TODO - is it ok for us to re-use schemas which is optional parameter for diff
```
Comment (zpanderson) [inline liquibase-core/src/main/java/liquibase/integration/commandline/Main.java:1771]: Wanted to verify if it's ok to re-use existing parameters from other commands.

Comment (molivasdat) [reply liquibase-core/src/main/java/liquibase/integration/commandline/Main.java:1771]: I'll have to take a look further if that is done elsewhere.  I don't think it is.

**Comment by molivasdat** [comment]
Hi @zpanderson Thanks for adding this enhancement request.  Can you provide the documentation on how to run the command?
Run dbDoc with multiple schemas in the optional parameter is pretty vague.

Here's what happens next:
A member of the Liquibase team will take a look at your contribution and may suggest: 

- Changes based on [guidelines for PR submission](https://liquibase.jira.com/wiki/spaces/LB/pages/1274904896/How+to+Contribute+Code+to+Liquibase+Core)
- Additional tests ([unit ](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)and/or [integration](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)) and/or [Test Harness Tests](https://github.com/liquibase/liquibase-test-harness/pulls)
- Provide other feedback, like requesting changes. 

The PR will be prioritized according to our internal development and testing capacity.

We’ll let you know when it’s ready to move to the next step or if any changes are needed.

File: liquibase-core/src/main/java/liquibase/integration/commandline/Main.java (hunk 1 above)
Comment (zpanderson) [reply liquibase-core/src/main/java/liquibase/integration/commandline/Main.java:1771]: Understandable, just figured if the code was already there to handle --schemas and the usage was the same that it'd be OK. If not, that's OK too, just more additions.

**Comment by zpanderson** [comment]
> 
> 
> Hi @zpanderson Thanks for adding this enhancement request. Can you provide the documentation on how to run the command?
> Run dbDoc with multiple schemas in the optional parameter is pretty vague.
> 
> Here's what happens next:
> A member of the Liquibase team will take a look at your contribution and may suggest:
> 
>     * Changes based on [guidelines for PR submission](https://liquibase.jira.com/wiki/spaces/LB/pages/1274904896/How+to+Contribute+Code+to+Liquibase+Core)
> 
>     * Additional tests ([unit ](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)and/or [integration](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)) and/or [Test Harness Tests](https://github.com/liquibase/liquibase-test-harness/pulls)
> 
>     * Provide other feedback, like requesting changes.
> 
> 
> The PR will be prioritized according to our internal development and testing capacity.
> 
> We’ll let you know when it’s ready to move to the next step or if any changes are needed.

In the existing use-case you run 'dbDoc <outputDir>' with no parameters and you still get the same behavior. If you run 'dbDoc <outputDir> --schemas=schema1,schema2' then you will have multiple schemas in the generated dbDoc. Or the schemas can be provided in the liquibase.properties file. I also added the optional parameter into the commandline help text. 

Is there a way to update the web docs (https://docs.liquibase.com/commands/community/dbdoc.html)? If so, I missed it so sorry on that.

File: liquibase-core/src/main/java/liquibase/integration/commandline/Main.java (hunk 1 above)
Comment (zpanderson) [reply liquibase-core/src/main/java/liquibase/integration/commandline/Main.java:1771]: @molivasdat - Any further thoughts on this?

**Comment by molivasdat** [comment]
Hi @zpanderson Thanks for the description on how to run the new enhancement.  We do have a way to add a PR to the docs and change the documentation.
It is located here [https://github.com/Datical/liquibase-docs](https://github.com/Datical/liquibase-docs)

File: liquibase-core/src/main/java/liquibase/integration/commandline/Main.java (hunk 1 above)
Comment (zpanderson) [reply liquibase-core/src/main/java/liquibase/integration/commandline/Main.java:1771]: @molivasdat - sorry to pester here, just would like to understand the next steps as to how to resolve this open question so we could proceed.

Comment (zpanderson) [reply liquibase-core/src/main/java/liquibase/integration/commandline/Main.java:1771]: @molivasdat - it's still not clear to me what the next steps are here. Any feedback you could provide would be appreciated.

**Comment by zpanderson** [comment]
@molivasdat - Anything more I need to do on this or just wait?

File: liquibase-core/src/main/java/liquibase/integration/commandline/Main.java (hunk 1 above)
Comment (nvoxland) [reply liquibase-core/src/main/java/liquibase/integration/commandline/Main.java]: Sorry for the slow responses. 

The --schemas usage is great

**Review by nvoxland — APPROVED** [review]
I resolved the merge conflicts due to changes in how we define our arguments. 

I also ended up changing the format of the table list on the left side to not use the collapsing schema but instead have more of a schema "header". That seemed to fit more in to the " '90s javadoc style vibe" we currently have going on, and we can look at a general nicer and more modern look and feel as something separate.

Otherwise, the changes were all good and we'll work on (finally) getting them merged into master for release. 

Code review and test results:

Things to be aware of:
- I did some manual testing and it worked well for me. 
- No automated tests, but not a good way to do automated testing at this point. So we'll want a bit more manual testing first
- Only impacts dbdoc, both with --schemas specified and not

Things to worry about:
- Nothing

**Review by suryaaki2 — APPROVED** [review]

**Review by filipelautert — APPROVED** [review]

**Comment by FBurguer** [comment]
Hello! I did some straight forward tests for this branch and found a behavior, that I dont think its incorrect, but I think its important to mention:
- `dbdoc` works with and without the schemas atttribute correctly, it snapshots all the schemas mentioned in the attribute, and if you dont add the attribute, it snapshots only the defaultSchema content (you can change the defaultSchema on the command call or the properties file).
- `dbdoc `doesnt warn or fails when you mention a non existing schema in the schemas attribute. 
```
liquibase dbdoc --output-directory=output --schemas=public,test,test2
####################################################
##   _     _             _ _                      ##
##  | |   (_)           (_) |                     ##
##  | |    _  __ _ _   _ _| |__   __ _ ___  ___   ##
##  | |   | |/ _` | | | | | '_ \ / _` / __|/ _ \  ##
##  | |___| | (_| | |_| | | |_) | (_| \__ \  __/  ##
##  \_____/_|\__, |\__,_|_|_.__/ \__,_|___/\___|  ##
##              | |                               ##
##              |_|                               ##
##                                                ##
##  Get documentation at docs.liquibase.com       ##
##  Get certified courses at learn.liquibase.com  ##
##  Free schema change activity reports at        ##
##      https://hub.liquibase.com                 ##
##                                                ##
####################################################
Starting Liquibase at 12:15:09 (version [Core: //dbDocs_Support_Multiple_Schemas/6177/4b8751/2022-12-22 16:51+0000, Pro: master/2601/89ad8e/2022-12-23T12:25:25Z] #6177 built at 2022-12-22 16:51+0000)
Liquibase Version: [Core: //dbDocs_Support_Multiple_Schemas/6177/4b8751/2022-12-22 16:51+0000, Pro: master/2601/89ad8e/2022-12-23T12:25:25Z]
Liquibase Pro [Core: //dbDocs_Support_Multiple_Schemas/6177/4b8751/2022-12-22 16:51+0000, Pro: master/2601/89ad8e/2022-12-23T12:25:25Z] by Liquibase licensed to Liquibase Pro Customer until dom abr 30 21:00:00 UYT 2023
Liquibase command 'dbDoc' was executed successfully.
```
image
I hope you found this helpful!

**Comment by filipelautert** [comment]
Hello @FBurguer - thanks for spotting this, I added a validation to the code. Can you confirm it is working as expected now?

File: liquibase-core/src/main/java/liquibase/changelog/visitor/DBDocVisitor.java [hunk 2]
```diff
@@ -170,6 +183,28 @@ public void writeHTML(Resource rootOutputDir, ResourceAccessor resourceAccessor)
...
+        Set<String> requiredSchemaNames = Arrays.stream(schemaList).filter(s -> !StringUtil.isEmpty(s.getSchemaName()))
+                .map(s -> s.getSchemaName().toLowerCase()).collect(Collectors.toSet());
+        List<String> notFoundSchemas = new ArrayList<>();
+
+        for (String required : requiredSchemaNames) {
+            if (!schemasNamesAtDb.contains(required)) {
+                notFoundSchemas.add(required);
+            }
+        }
+        if (!notFoundSchemas.isEmpty()) {
```
Comment (XDelphiGrl) [inline liquibase-core/src/main/java/liquibase/changelog/visitor/DBDocVisitor.java:203]: Thank you for adding the validation suggested by @FBurguer, @zpanderson!

File: liquibase-core/src/main/java/liquibase/changelog/visitor/DBDocVisitor.java [hunk 3]
```diff
@@ -123,7 +128,15 @@ public void writeHTML(Resource rootOutputDir, ResourceAccessor resourceAccessor)
...
         copyFile("liquibase/dbdoc/overview-summary.html", rootOutputDir);
 
-        DatabaseSnapshot snapshot = SnapshotGeneratorFactory.getInstance().createSnapshot(database.getDefaultSchema(), database, new SnapshotControl(database));
+        CatalogAndSchema[] computedSchemaList = schemaList;
+        if (schemaList == null) {
+            computedSchemaList = new CatalogAndSchema[]{database.getDefaultSchema()};
+        }
+
+        DatabaseSnapshot snapshot = SnapshotGeneratorFactory.getInstance().createSnapshot(computedSchemaList, database, new SnapshotControl(database));
+        if (schemaList != null && schemaList.length != 0 && !(database instanceof FirebirdDatabase)) {
```
Comment (XDelphiGrl) [inline liquibase-core/src/main/java/liquibase/changelog/visitor/DBDocVisitor.java:137]: May I inquire as to why there is special handling for the Firebird database?

**Comment by XDelphiGrl** [comment]
I will validate the edge case @FBurguer discovered is addressed by the changes to the PR. I will also add an internal Liquibase ticket to create functional tests for the `dbDoc` command.

CC @zpanderson

File: liquibase-core/src/main/java/liquibase/changelog/visitor/DBDocVisitor.java (hunk 3 above)
Comment (filipelautert) [reply liquibase-core/src/main/java/liquibase/changelog/visitor/DBDocVisitor.java:137]: Firebird has some strange behaviour:
* It doesn't support schemas
* But in order to the createSnapshot to work for it, it requires a schema name of "default"
* But (again) when we do " snapshot.get(Schema.class);" in the validation, it returns no schemas, causing the validation to fail.

Besides we did a lot of stuff for Firebird recently it still buggy (a lot of tests do no work for it). So I decided to just exclude him from this validation instead of disabling dbDoc tests. If one day we fix it we can remove this condition (as in many other places).

Comment (XDelphiGrl) [reply liquibase-core/src/main/java/liquibase/changelog/visitor/DBDocVisitor.java:137]: Thanks for the explanation; your decision makes sense to me.

**Comment by XDelphiGrl** [comment]
@filipelautert and @zpanderson, hi! 

Overall, everything works well. In testing @FBurguer's test case, I found that Liquibase `dbDoc` correctly warns when a specified schema cannot be located. If one schema is specified, and one schema not found, the error message is:
 
```
Could not find any of the required schemas at the configured database.
```
If a list of schemas is provided and one in the list cannot be found, the error message is:

```
The following schema(s) could not be found at database: x
```
However, when there is a missing schema, the output-directory is still created. The output-directory contains the globalnav, index and overview-summary HTML files. This results in broken links in the documentation. It would be better not to create the directory at all when there is an error finding a schema to document. What do you think? Should there be a separate issue created, or is this a minor enough change to address in this PR?

image

**Comment by filipelautert** [comment]
@XDelphiGrl fixed! (I hope so)

**Comment by XDelphiGrl** [comment]
@filipelautert , hi!

Thanks for the fix to prevent creation of incomplete documentation when Liquibase cannot locate a schema to document. It works as expected.

---
**Test Environment**
Build: [https://github.com/liquibase/liquibase/actions/runs/3913991635](https://github.com/liquibase/liquibase/actions/runs/3913991635)
OSS: dbDocs_Support_Multiple_Schemas/6602/57affd/
Commercial: master/2663/5d2338/
Postgres 12.6

**Review by XDelphiGrl — APPROVED** [review]
The `db-doc` command now supports generating schema documentation for multiple schemas! This is accomplished by specifying the `--schemas` parameter on your `db-doc` command. If a schema cannot be located, an error is thrown instead of creating partial documentation in the `--output-directory`.

**Example Command Specifying --schemas**
```
liquibase db-doc --changelog-file communityOnly.xml --schemas public --output-directory dbDocs-pg-public
```

* Automated functional and test harness executions passing.
* No additional testing required.

APPROVED
